### PR TITLE
feat(cli): :sparkles: Add option for defining theme name

### LINKED ISSUE #2334: Add support for passing name to create tokens script

### PR DESCRIPTION
**Pull request by mimarz**
resolves #2334

- Added new option to `--theme` to cli for defining theme name.
  - Will add new theme or overwrite existing theme.
- Added new script in cli (`update:template`) for updating template files from our "master" `design-tokens` in root.
- Added textfield in theme-builder tokens modal for defining theme name.
- Not a fan of having code, but commented our template generation for `$metadata` and `$themes` as this is an internal improvement that can be done later.
- Some updates to components used in theme-builder to get desired user interface.

### CODE REVIEW

File: packages/cli/bin/designsystemet.ts
Comment (unekinn) [inline packages/cli/bin/designsystemet.ts:46]: I think, if you change line 41 to `.option('-f, --font-family <string>', `Font family`, 'Inter')`, you can remove the conditional, as the Commander option type is then no longer `boolean|string`, and the default value is `'Inter'`.

File: packages/cli/bin/designsystemet.ts
Comment (unekinn) [inline packages/cli/bin/designsystemet.ts:40]: Maybe define the color options as `requiredOption` instead of `option` while you're at it?

File: packages/cli/src/tokens/template.ts
Comment (unekinn) [inline packages/cli/src/tokens/template.ts:5]: Remove or keep, depending on decision on the big commented code block below

File: packages/cli/src/tokens/template.ts
Comment (unekinn) [inline packages/cli/src/tokens/template.ts:19]: Remove or keep, depending on decision on the big commented code block below

File: packages/cli/src/tokens/template.ts
Comment (unekinn) [inline packages/cli/src/tokens/template.ts:36]: Not a big fan of keeping code in a comment. I would prefer it to be a function in a separate file, but if you think it will be used VERY soon I guess it's fine.

File: packages/cli/src/tokens/template.ts
Comment (unekinn) [inline packages/cli/src/tokens/template.ts:94]: Is this run as a side effect on purpose? Why?

File: packages/cli/src/tokens/write.ts
Comment (unekinn) [inline packages/cli/src/tokens/write.ts:15]: `\r\n` (crlf) smells like missing line ending normalization in the repository, but `.gitattributes` has
```
* text=auto eol=lf
```
So does this actually do anything?

**Review by unekinn — COMMENTED** [review]
Some preliminary comments, I'll look at `src/tokens/write` and test locally after lunch!

File: apps/theme/components/TokenModal/TokenModal.tsx
Comment (mimarz) [reply apps/theme/components/TokenModal/TokenModal.tsx:101]: Changed to to "Navn på tema"

File: packages/cli/bin/designsystemet.ts
Comment (mimarz) [reply packages/cli/bin/designsystemet.ts:46]: When you have an "optional option-argument" (defined by brackets`[string]`), the type is still `string|true` even if you have a default defined.

But thats maybe the wrong option type to to use so I'll change it to "required option-argument"

File: packages/cli/bin/designsystemet.ts
Comment (mimarz) [reply packages/cli/bin/designsystemet.ts:48]: fixed

File: packages/cli/bin/designsystemet.ts
Comment (mimarz) [reply packages/cli/bin/designsystemet.ts:40]: Good point

File: packages/cli/bin/designsystemet.ts
Comment (unekinn) [reply packages/cli/bin/designsystemet.ts:46]: This is incorrect. According to the docs, `[string]` means `boolean|string` and `<string>` means "just" `string`. It has nothing to do with optional or required, which is defined through `.option` or `requiredOption` respectively

**Review by unekinn — APPROVED** [review]
The script seems to work well. No unexpected changes in the tokens compared to previous versions, and adding a second theme also works. Good job 🥇

I had a couple of questions in the earlier review which I'd like answered, but feel free to merge and potentially fix the rest later...

File: packages/cli/bin/designsystemet.ts
Comment (mimarz) [reply packages/cli/bin/designsystemet.ts]: It should be correct now :)

I wrote option type in the last sentence, but was supposed to be argument type.

Only `--write` is "optional option-argument" now. ie, you can use it as `--write` or with argument `--write "./my-design-tokens"`

Docs is clear on this once you understand the difference between option and option argument 😅 
image

File: packages/cli/src/tokens/template.ts
Comment (mimarz) [reply packages/cli/src/tokens/template.ts:94]: I only use it as a npm-script for now so thats why.

`"update:template": "tsx ./src/tokens/template.ts"`

File: packages/cli/src/tokens/write.ts
Comment (mimarz) [reply packages/cli/src/tokens/write.ts:15]: I had to remove the line ending from the design-tokens files because Token Studio strips them once imported, so a lot of files was flagged as changed the first time you use them.

Comment (unekinn) [reply packages/cli/src/tokens/write.ts:15]: That makes sense, but I would think it should be `.replace(/\n/g, '')`. On my system, there are no files in the repo which have `\r\n` in them, and `eol=lf` in `.gitattributes` SHOULD force Git to normalize line endings to `\n` on all systems on checkout, so I don't understand how this actually does any replacement of line endings.

Alternatively `/\r?\n/g` should work regardless of system (only Windows uses `\r\n` for line endings by default) but it's weird if you have `\r\n` in any files now with the current `.gitattributes` file

Comment (mimarz) [reply packages/cli/src/tokens/write.ts:15]: I tried `.replace(/\n/g, '')`first time but it removes all the new lines after each prop so file ended up just gibberish. Same with `/\r?\n/g`. Maybe its due to the spacer parameter to `JSON.stringify`. 

image


I did some trial and error before landing on the one used now that works. 🤷

File: packages/cli/src/tokens/write.ts
Comment (mimarz) [reply packages/cli/src/tokens/write.ts:65]: We are exclusively looking for `$themes.json` here so figured it would be ok. The alternative is to use [fs.access](https://nodejs.org/api/fs.html#fs_fs_access_path_mode_callback), but I figured this would be ok for now 😅

File: packages/cli/src/tokens/write.ts
Comment (unekinn) [reply packages/cli/src/tokens/write.ts:15]: Well, on my end the script works the same way with and without `.replace(/\r\n/g, '');` so I don't think that particular invocation does anything, unless you have added a Windows line feed to your file in an editor which doesn't detect existing line endings, and not commited it

Comment (unekinn) [reply packages/cli/src/tokens/write.ts:15]: To be clear I think the line should just be
```
export const stringify = (data: unknown) => JSON.stringify(data, null, 2)
```

 since this already strips the newline before the end of the file

File: packages/cli/src/tokens/write.ts
Comment (unekinn) [reply packages/cli/src/tokens/write.ts:65]: Yeah if you don't have read access to $themes.json you probably don't have write access either (although it's theoretically possible)

File: packages/cli/src/tokens/write.ts
Comment (mimarz) [reply packages/cli/src/tokens/write.ts:15]: ahh, true. Same here. I'll just remove it then and we can add it later if its needed again.

**Pull request merged**
